src/back/glsl/features.rs

```diff
@@ -33,6 +33,8 @@ bitflags::bitflags! {
         /// Arrays with a dynamic length
         const DYNAMIC_ARRAY_SIZE = 1 << 16;
         const MULTI_VIEW = 1 << 17;
+        /// Adds support for fused multiply-add
+        const FMA = 1 << 18;
```

Comment:
Why do we need thus flag? If FMA isn't natively supported, we are emulating it anyway. So it seems to me that this flag isn't getting us anything.

Reply:
For one thing, I'm following precedent with existing features that need extensions, and `fma` is only supported on GLES 3.1+ with: https://github.com/gfx-rs/naga/blob/7454d1fc04cd3b74474016a56e870b640897e1eb/tests/out/glsl/functions.main.Compute.glsl#L1-L2

This PR does not emulate `fma` on GLSL in every case, but decides if it must emulate it or else it requests the extension when necessary. This fixes that unusual validation error you noted earlier: https://github.com/gfx-rs/naga/runs/4446174291?check_suite_focus=true

```
ERROR: 0:13: 'fma' : required extension not requested: Possible extensions include:
GL_EXT_gpu_shader5
GL_OES_gpu_shader5
```

I chose to use the existing feature flag infrastructure to write this extension, rather than coming up with something unique just for this case. Is there something better I could have done here?

The `FMA` feature flag name is probably too narrow, honestly. `GL_EXT_gpu_shader5` enables a lot more than just the `fma` function, and the feature flag can be used to support all of it on GLES.

Reply:
The capability flags in GLSL backend are meant to be requirements. I.e. shader requires A, B, C, and we want to check if we can work with this shader at all.
The case for FMA is different. The backend *always* supports FMA instruction. The only thing different is a code path taken. Therefore, there is no case where GLSL backend would check for this capability and report it missing. It's not a real capability.

Reply:
I think I understand what you mean. Let me try to rephrase it; The `fma` function from the frontend is always supported by the backend (GLSL) even if it has to fallback to an arithmetic transformation (i.e., "emulated"). This PR uses a feature flag (capability) in another sense, that it can enable the use of the GLSL `fma` function on particular versions of the backend. Which are not how feature flags are used elsewhere.

Would that be an accurate way to describe the situation?

I'll have to think on it if I need to use some other mechanism to enable the extension for GLES. I do see an extension enabled that is not controlled by feature flags: https://github.com/gfx-rs/naga/blob/7c8bedcb1c6f9bb14b4e9fc8f48496bea3a67491/src/back/glsl/mod.rs#L442-L450 I suppose that would be an appropriate place to put this?